File: Dokie/ViewControllers/Login/LoginViewModel.m
//
//  LoginViewModel.m
//  Doci
//
//  Created by Ashamaz Shidov on 02/12/2018.
//  Copyright © 2018 Ashamaz Shidov. All rights reserved.
//

#import "LoginViewModel.h"
#import "NavigationRouter.h"
#import "EncryptionManager.h"
#import "AppDelegate.h"
#import "LocalAuth.h"

@interface LoginViewModel ()

@property (nonatomic, strong) RACSubject *enableInputSubject;
@property (nonatomic, strong) NSString *loginLabelText;
@property (nonatomic, strong) NSString *loginButtonText;
@property (nonatomic) BOOL touchIdLoginEnabled;
@property (nonatomic) BOOL faceIdLoginEnabled;

@property (nonatomic, strong, readonly) LocalAuth *localAuth;
@property (nonatomic, strong, readonly) NavigationRouter *navigationRouter;
@property (nonatomic, strong, readonly) NSManagedObjectContext *objectContext;

@end

@implementation LoginViewModel

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    RAC(self, touchIdLoginEnabled) = [RACObserve(self, localAuth.touchIdAvailable) ignore:nil];
    RAC(self, faceIdLoginEnabled) = [RACObserve(self, localAuth.faceIdAvailable) ignore:nil];
    self.enableInputSubject = [RACSubject new];
    [self updateText];
}

- (void)updateText {
    self.loginLabelText = [EncryptionManager contextHasKey:self.objectContext] ? NSLocalizedString(@"Password:", @"Password label text") : NSLocalizedString(@"Welcome to Dokie!\nCreate password for your database:", @"Create password label text");
    self.loginButtonText = [EncryptionManager contextHasKey:self.objectContext] ? NSLocalizedString(@"Login", @"Login button text") : NSLocalizedString(@"Create database", @"Create database button text");
    if (![EncryptionManager contextHasKey:self.objectContext]) {
        [self.localAuth resetStoredPassword];
    }
}

- (void)loginPressedWithPassword:(NSString *)password {
    NSString *loadingGuid = [self.navigationRouter showLoading];
    @weakify(self);
    dispatch_async(dispatch_get_main_queue(), ^{
        @strongify(self);
        EncryptionManager *encryptionManager = [[EncryptionManager alloc] initWithPassword:password objectContext:self.objectContext];
        if ([encryptionManager isValid]) {
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter pushMainMenuWithEncryptionManager:encryptionManager managedObjectContext:self.objectContext];
            dispatch_async(dispatch_get_main_queue(), ^{
                @strongify(self);
                [self.localAuth validPasswordEntered:password];
            });
        }
        else {
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Invalid password", @"Invalid password message")];
        }
        [self.enableInputSubject sendNext:nil];
    });
}

- (void)showAbout {
    [self.navigationRouter showAbout];
}

- (void)loginWithBiometrics {
    @weakify(self);
    [[[self.localAuth databasePassword]
      deliverOnMainThread]
     subscribeNext:^(NSString *password) {
        @strongify(self);
        [self loginPressedWithPassword:password];
    }
     error:^(NSError * _Nullable error) {
        @strongify(self);
        [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Can not retrieve password from keychain", @"Keychain error")];
    }];
}

#pragma mark -

- (NavigationRouter *)navigationRouter {
    return [NavigationRouter shared];
}

- (LocalAuth *)localAuth {
    return [LocalAuth shared];
}

- (NSManagedObjectContext *)objectContext {
    AppDelegate *delegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    return delegate.objectContext;
}

@end

File: Dokie/ViewControllers/Login/LoginViewModel.swift
//
//  LoginViewModel.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 14.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import Foundation
import ReactiveCocoa
import ReactiveSwift
import CoreData

class LoginViewModel : NSObject {
    
    let enableInput = MutableProperty<Bool>(true)
    let loginLabelText = MutableProperty<String>("")
    let loginButtonText = MutableProperty<String>("")
    let touchIdLoginEnabled = MutableProperty<Bool>(false)
    let faceIdLoginEnabled = MutableProperty<Bool>(false)
    var biometricsAvailable: Bool {
        touchIdLoginEnabled.value || faceIdLoginEnabled.value
    }
    private let localAuth = LocalAuth.shared
    private let objectContext: NSManagedObjectContext!
    private let navigationRouter = NavigationRouter.shared
    
    override init() {
        if let appDelegate = UIApplication.shared.delegate as? AppDelegate {
            self.objectContext = appDelegate.objectContext
        }
        else {
            self.objectContext = nil
        }
        super.init()
        setup()
    }
    
    private func setup() {
        touchIdLoginEnabled <~ localAuth.touchIdAvailable
        faceIdLoginEnabled <~ localAuth.faceIdAvailable
        updateText()
    }
    
    func loginPressed(password: String?) {
        guard let password = password else {
            navigationRouter.showAlert(title: "", message: String.localized("Invalid password"))
            return
        }
        
        enableInput.value = false
        let loadingGuid = navigationRouter.showLoading()
        DispatchQueue.main.async { [unowned self] in
            let encryptionManager = EncryptionManager(password: password, context: self.objectContext)
            self.navigationRouter.hideLoading(loadingGuid)
            if encryptionManager.isValid {
                self.navigationRouter.pushMainMenu(encryptionManager: encryptionManager, managedObjectContext: self.objectContext)
                DispatchQueue.main.async { [unowned self] in
                    self.localAuth.validPasswordEntered(password)
                }
            }
            else {
                self.navigationRouter.showAlert(title: "", message: String.localized("Invalid password"))
            }
            self.enableInput.value = true
        }
    }
    
    func loginWithBiometrics() {
        localAuth
            .databasePassword()
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case let .success(password):
                    self.loginPressed(password: password)
                default:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Can not retrieve password from keychain"))
                }
        }
    }
    
    func showAbout() {
        navigationRouter.showAbout()
    }
    
    func updateText() {
        let hasKey = EncryptionManager.keyExists(in: objectContext)
        loginLabelText.value = hasKey ? String.localized("Password:") : String.localized("Welcome to Dokie!\nCreate password for your database:")
        loginButtonText.value = hasKey ? String.localized("Login") : String.localized("Create database")
        if hasKey == false {
            localAuth.resetStoredPassword()
        }
    }
    
}
